Fixed DateAndSequence archive numbering mode + bugfix no max archives

remye06 opened the pull request:
This PR addresses various issues inherent to the DateAndSequence archive numbering mode :
- The Sequence part wouldn't be padded accordingly ;
- MaxArchiveFiles wouldn't be respected under some circumstances ;
- Using any standard format as ArchiveDateFormat would lead to an ArgumentOutOfRangeException being thrown from a call to String.Substring ;
- Performance improvement by using Int32.TryParse instead of Convert.ToInt32 and exception handling as flow control

304NotModified commented:
This looks good, but can you add unit test(s) for this? (for bullet 1 and 3)

304NotModified commented:
@remye06 do you know if this will fix #643?

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -996,13 +994,23 @@ private void DateAndSequentialArchive(string fileName, string pattern, LogEventI
+                    .Where(n => n.Name.Length >= archiveFileNameMinLength)
+                    .OrderBy(n => n.CreationTime)
+                    .Select(n => n.FullName)
+                    .ToList();
```
We could remove some duplicate code here, by adding an intermediate result  (variable of type `ienumerable<string>`.

the duplicate code:
```
                .OrderBy(n => n.CreationTime)
                   .Select(n => n.FullName)
                    .ToList();

```

could you remove the duplicate code?

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -1023,18 +1031,15 @@ private void DateAndSequentialArchive(string fileName, string pattern, LogEventI
-                        num = Convert.ToInt32(numberPart, CultureInfo.InvariantCulture);
-                    }
-                    catch (FormatException)
+                    if (!Int32.TryParse(numberPart, NumberStyles.None, CultureInfo.InvariantCulture, out num))
```
can you add some comments of what we're trying to archive over here? (same for `DateTime.TryParseExact` below). 

Before fully understanding this code, I think it should be a separate method (`detectArchiveFile` or something?) - correct me if wrong.

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -967,11 +967,9 @@ private void DateAndSequentialArchive(string fileName, string pattern, LogEventI
-            int dateTrailerLength = baseNamePattern.Length - placeholderLastPart;
-
-            string fileNameMask = baseNamePattern.Substring(0, placeholderFirstPart) + "*" + baseNamePattern.Substring(placeholderLastPart);
+            FileNameTemplate fileTemplate = new FileNameTemplate(baseNamePattern);
```
please comment there lines of what we're trying to archive.

304NotModified commented:
Your code looks solid, but I think we could improve something in code understanding (comments, methods). Do you agree? See notes.

remye06 commented:
Hi, 
I added a test for the padding.
I factored the file enumeration as you asked to remove duplicate code.
I also found a new bug that I fixed on the fly : the behavior with no MaxArchive (0, the default) was different from the other archive modes. 1 archive would always be replaced instead of an unlimited number of archives being kept.

Further commits are on their way to fulfill the rest of your comments.

304NotModified commented:
Thanks, let us know when it's completed :)

One question, the bug you have found and solved, this was a bug already in the 4.0 release?

304NotModified commented:
PS you can ignore this unit test. It's a race condition issue. But difficult to solve, because it only fails 5% of the times.

remye06 commented:
Yep, I didn’t introduce it myself if that’s what you mean :p

De : Julian Verdurmen [mailto:notifications@github.com]
Envoyé : mercredi 1 juillet 2015 20:48
À : NLog/NLog
Cc : Rémy Esmery
Objet : Re: [NLog] Fixed DateAndSequence archive numbering mode (#773)


Thanks, let us know when it's completed :)

One question, the bug you have found and solved, this was a bug already in the 4.0 release?

—
Reply to this email directly or view it on GitHub<https://github.com/NLog/NLog/pull/773#issuecomment-117793384>.

304NotModified commented:
I would like to merge this, but IMO I needs some more comments in the code. Can you fix that?

remye06 commented:
Yep I'll do that ASAP

Sent from my Sony Xperia™ smartphone


---- Julian Verdurmen wrote ----


I would like to merge this, but IMO I needs some more comments in the code. Can you fix that?

—
Reply to this email directly or view it on GitHub<https://github.com/NLog/NLog/pull/773#issuecomment-122441296>.

304NotModified commented:
Thanks in advance!

304NotModified commented:
note to myself: ignore failed unit test

remye06 commented:
I don't think this will fix #643 since it's a completely different codepath.

304NotModified commented:
Thanks for the feedback!

304NotModified commented:
I will merge this when it has some more comments are mentioned before.

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -994,86 +991,125 @@ private void DateAndSequentialArchive(string fileName, string pattern, LogEventI
 
-            int nextSequenceNumber = -1;
-
+            int minSequenceLength = fileTemplate.EndAt - fileTemplate.BeginAt - 2;
```
add comment (in c# code, max one line) - what are we trying to archive here.

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -994,86 +991,125 @@ private void DateAndSequentialArchive(string fileName, string pattern, LogEventI
-                var archiveFileCount = files.Count; 
+                var currentDateArchives = archives
+                    .Where(a => a.HasSameArchiveDate(archiveDate))
+                    .ToArray();
```
Please use toList

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -994,86 +991,125 @@ private void DateAndSequentialArchive(string fileName, string pattern, LogEventI
+            }
+            catch (DirectoryNotFoundException)
+            {
+                Directory.CreateDirectory(dirName);
```
add internal logger (info)

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -994,86 +991,125 @@ private void DateAndSequentialArchive(string fileName, string pattern, LogEventI
-                        archiveFileCount--;
-                        continue;
-                    }
+            string paddedSequence = nextSequenceNumber.ToString().PadLeft(minSequenceLength, '0');
```
add comment (in c# code, max one line) - what are we trying to archive here.

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -994,86 +991,125 @@ private void DateAndSequentialArchive(string fileName, string pattern, LogEventI
+        /// </summary>
+        /// <remarks>No file is deleted if MaxArchiveFile is equal to zero.</remarks>
+        private void EnsureAvailableArchiveFile(DateAndSequenceArchive[] archives)
+        {
```
Please List instead of Array

304NotModified commented:
Code looks nice! But there are some small improvements needed:

- No Arrays please, but lists. (and dictionaries/hashsets when using .Contains)
- Some more documentation is needed. Please document WHY we are doing this, not HOW. 
- even internal classes should be fully documented. I have seen too many bugs in NLog because of missing docs...(and ambiguous names)

Addi-B commented:
Just noticed another error. When using the following code the MaxArchiveFiles function starts to delete the files starting from sequence number 10 instead of continuing to delete the single digit numbers if it didn't delete all single digit numbers before midnight.

    fileTarget.ArchiveFileName = filePath + LogName  +".{##}.log";
    fileTarget.KeepFileOpen = false;
    fileTarget.ArchiveNumbering = ArchiveNumberingMode.DateAndSequence;
    fileTarget.ArchiveEvery = FileArchivePeriod.Day;
    fileTarget.MaxArchiveFiles = 30;
    fileTarget.ArchiveAboveSize = 1048576; // 1 MB
    fileTarget.Layout = "${longdate}|${message}"; 

The resulting logfile names are:

    something.20150812.0.log
    ...
    something.20150812.9.log
    something.20150812.10.log
    ...

304NotModified commented:
Hi, 

The last noticed bug can be fixed in another PR?

304NotModified commented:
(maybe we should create a new issue for that, review this and merge it)

Addi-B commented:
I'll go ahead and create and create a new issue.

304NotModified commented:
Thanks!

304NotModified commented:
Can you handle the code comments?

remye06 commented:
Hi, will do next week, I'm on holidays atm


---- Julian Verdurmen wrote ----


Can you handle the code comments?

-
Reply to this email directly or view it on GitHub<https://github.com/NLog/NLog/pull/773#issuecomment-131182261>.

304NotModified commented:
OK, I will check it out. Have a good holiday!

304NotModified commented:
I will handle the unit test for #847. Have a good holiday

remye06 commented:
Working on it atm, I messed up while rebasing. That will be fixed by the day.

304NotModified commented:
Thanks for the up

remye06 commented:
Just a few reviews of yours to handle (mainly comments) and we're good to go !

304NotModified commented:
OK, let me know when it's done :)

remye06 commented:
I think we're done here, can you confirm that all is well ?

304NotModified commented:
I will do that this week.

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -1178,13 +1185,17 @@ private void DeleteOldDateArchive(string pattern)
 
+                /* TODO: The following block could use EnsureArchiveCount, but the behavior is not exactly the same.
+                 * The number of files to delete to "make room" is calculated using 'files.Count' instead of 'filesByDate.Count',
+                 * which I suspect to be a bug, since 'files' can contain filenames that are not actual archives. */
```
This is a bug. It should by filebydate.  Van you fix that

remye06 replied:
I'd like to create a dedicated PR for this as soon as this one gets merged. Don't you prefer ?

304NotModified replied:
That's also nice. :)

304NotModified commented on src/NLog/Targets/DateAndSequenceArchive.cs:
```diff
@@ -0,0 +1,98 @@
+        private readonly DateTime _date;
+        private readonly string _dateFormat;
+        private readonly int _sequence;
+        private readonly string _formattedDate;
```
If you have time : convert to readonly auto properties

304NotModified commented:
Checked it already! 2 small things and then I will merge it. The refactoring looks very good! :+1:

304NotModified commented:
I'm impressed by your code and style! 

Aren't you interested in http://nlog-project.org/2015/08/05/NLog-is-looking-for-developers.html

304NotModified commented:
Merged :)
